core/src/es/adapter/transformer/strategy/mod.rs

```diff
@@ -0,0 +1,99 @@
+//! [`Strategy`] definition and default implementations.
+
+pub mod as_is;
+pub mod custom;
+pub mod initialized;
+pub mod into;
+pub mod skip;
+pub mod split;
+
+use futures::Stream;
+
+use crate::es::{adapter, event};
+
+use super::{Transformer, WithStrategy};
+
+#[doc(inline)]
+pub use self::{
+    as_is::AsIs,
+    custom::{Custom, Customize},
+    initialized::Initialized,
+    into::Into,
+    skip::Skip,
+    split::{Split, Splitter},
+};
+
+/// Generalized [`Transformer`] for [`Versioned`] events.
+///
+/// [`Versioned`]: event::Versioned
+pub trait Strategy<Adapter, Event, Ctx>
+where
+    Ctx: ?Sized,
+{
+    /// Error of this [`Strategy`].
+    type Error;
+
+    /// Converted [`Event`].
+    ///
+    /// [`Event`]: crate::es::Event
+    type Transformed;
+
+    /// [`Stream`] of [`Transformed`] [`Event`]s.
+    ///
+    /// [`Event`]: crate::es::Event
+    /// [`Transformed`]: Self::Transformed
+    type TransformedStream<'out>: Stream<
+            Item = Result<
+                <Self as Strategy<Adapter, Event, Ctx>>::Transformed,
+                <Self as Strategy<Adapter, Event, Ctx>>::Error,
+            >,
+        > + 'out;
+
+    /// Converts incoming [`Event`] into [`Transformed`].
+    ///
+    /// [`Event`]: crate::es::Event
+    /// [`Transformed`]: Self::Transformed
+    fn transform<'me, 'ctx, 'out>(
+        adapter: &'me Adapter,
+        event: Event,
+        context: &'ctx Ctx,
+    ) -> Self::TransformedStream<'out>
+    where
+        'me: 'out,
+        'ctx: 'out;
```

Comment:
@ilslv shouldn't it be opposite?

Just substitute `'static` for verifying:  
If we return a `'static` `Stream` then it demands `'me: 'static` and `'ctx: 'static`, which shouldn't be the case.

Correct bounds here are:
```rust
    where 'out: 'me + 'ctx
```

So we ensure that the `Stream` doesn't outlive its referees, and not that the referees don't outlive the `Stream`.

Reply:
@tyranron `where 'out: 'me + 'ctx` imply that `TransformedStream` has to outlive __longest__ of `'me` and `'ctx`. We want to bound to the __shortest__ of them. Initially I've implemented like that with expectation that compiler will shorten one of the lifetimes to satisfy bounds, but it didn't.

We have to basically perform `async fn` desugaring: https://github.com/rust-lang/rust/issues/56238#issuecomment-441736818

Reply:
@ilslv yup, you're right.